implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/healthcheck.ex

```diff
@@ -9,35 +9,38 @@ defmodule Ockam.Healthcheck do
 
   require Logger
 
-  def check_node(node_host, node_port, api_worker, ping_worker, timeout \\ 5000) do
-    node_address = Ockam.Transport.TCPAddress.new(node_host, node_port)
-    api_route = [node_address, api_worker]
-    {:ok, identity} = get_healthcheck_identity()
+  def check_targets() do
+    targets = Application.get_env(:ockam_healthcheck, :targets, [])
+    ## TODO: parallel check for multiple targets
+    Enum.each(targets, &Ockam.Healthcheck.check_target/1)
+  end
+
+  def check_target(
+        target,
+        timeout \\ 5000
+      ) do
     start_time = System.os_time(:millisecond)
 
-    case ping_node(api_route, identity, ping_worker, timeout) do
+    case ping_target(target, timeout) do
       :ok ->
-        report_check_ok(node_host, node_port, api_worker, ping_worker, start_time)
+        report_check_ok(target, start_time)
         :ok
 
       {:error, reason} ->
-        report_check_failed(node_host, node_port, api_worker, ping_worker, reason, start_time)
+        report_check_failed(target, reason, start_time)
         {:error, reason}
     end
   end
 
-  def ping_node(api_route, identity, worker, timeout) do
-    ## Improve error reporting for healthcheck
-    ## report secure channel and worker errors differently
-    case SecureChannel.create_channel(
-           route: api_route,
-           identity: identity,
-           key_exchange_timeout: 10_000
-         ) do
-      {:ok, channel} ->
+  def ping_target(target, timeout) do
+    %{host: host, port: port, api_worker: api_worker, healthcheck_worker: healthcheck_worker} =
+      target
+
+    with_tcp(host, port, fn conn ->
+      with_channel(conn, api_worker, fn channel ->
```

Comment:
Just a note, as I don't have a suggestion on how to simplify this, but it's difficult to follow these wrappers around anonymous functions nested inside each other. Probably a refactoring opportunity for the future (maybe by making it more explicit at a cost of it being less DRY?).

Reply:
Explicit wrapping of try/catch blocks doesn't look good at all and we don't really have a good resource cleanup RN. It's less about DRY as it is about readability.